### ISSUE TITLE
Noisy output

### ISSUE THREAD

**Issue by Nikronic** [Nikronic/CoarseNet#5]
I tried to overfit model on a simple and small dataset, but I got very noisy output. I tried to change target to edge detection using Canny edges as ground truth, but I sill have noisy output.

I am stuck right now and I tried to debug using TensorboardX but it has issues (I made one in their repository).
So plan is to implement my model using Tensorflow and try to debug it using tensorboard and convert the final model to Pytorch.

- [x] Implement model using Tensorflow
- [ ] Convert back model to PyTorch
- [x] Debug using tensorboard

Note: I also tried Visdom by FacebookResearch but they do not provide any API to demonstrate graph states.

**Comment by Nikronic** [comment]
Tensorboad graph visualization is same as the provided shape in U-Net paper.
Gradients are not checked yet.

**Comment by Nikronic** [comment]
`RandomNoise` layer is making images too noisy! Here is an example:

![out_noise_normalization](https://user-images.githubusercontent.com/32401852/59147909-c9bc8680-8a16-11e9-99a9-9917a1be3119.png)

---
`Normalization` makes images noisy a little bit. Here is an example:

![out_normalization](https://user-images.githubusercontent.com/32401852/59147930-038d8d00-8a17-11e9-8b29-a4b6df9a6655.png)

---
And here is an example without normalization and random noise:

![out-simple](https://user-images.githubusercontent.com/32401852/59147936-14d69980-8a17-11e9-8220-d2cd21dc8b49.png)

**Comment by Nikronic** [comment]
Here are trained outputs with default parameters for 20 epochs:
Without `Normalization` and `RandomNoise`:

![out-simple](https://user-images.githubusercontent.com/32401852/59147971-8282c580-8a17-11e9-8735-f15f7ca79077.png)
---
Without `Normalization` and with `RandomNoise`:

![out_normalization](https://user-images.githubusercontent.com/32401852/59147995-bc53cc00-8a17-11e9-813b-047c1646015b.png)
---

With `Normalization` and `RandomNoise`:

![out_noise_normalization](https://user-images.githubusercontent.com/32401852/59148005-d4c3e680-8a17-11e9-94df-ad13669902dc.png)

**Comment by Nikronic** [comment]
Note: Halftone effects applied randomly, so with every run we get different images.

**Comment by Nikronic** [comment]
`RandomNoise` implementation is wrong. Check implementation in `DeepHalftone` repository.'
Please see commit https://github.com/Nikronic/CoarseNet/commit/2be8962eddccfb35ef9c04589f67958777753651

**Comment by Nikronic** [comment]
`std = 0.007` has been chosen `RandomNoise` class. You can achieve it by  `0.1/128` division. `0.1` is the value proposed in the paper and `128` is the normalization term. Actually, `0.1` is a `std` value for images in range `[0, 255]`, but in Pytorch, images are between `[0,1]` and then we standardizing them between `[-1, 1]`, So we have to divide it by `128`.

**Comment by Nikronic** [comment]
Normalization is the reason we see creepy images even after loading images from the dataset, So the outputs should be more creepy!

You can see the abstract solution in this post: https://discuss.pytorch.org/t/understanding-transform-normalize/21730/18
And the similar question posted here: https://discuss.pytorch.org/t/understanding-transform-normalize/21730/17

**Comment by Nikronic** [comment]
> `std = 0.007` has been chosen `RandomNoise` class. You can achieve it by `0.1/128` division. `0.1` is the value proposed in the paper and `128` is the normalization term. Actually, `0.1` is a `std` value for images in range `[0, 255]`, but in Pytorch, images are between `[0,1]` and then we standardizing them between `[-1, 1]`, So we have to divide it by `128`.

`std=0.1` is THE CORRECT ONE.
Unnormalization can solve the issue.

**Comment by Nikronic** [comment]
After unnormalizing, here is the sample output:

![unnorm](https://user-images.githubusercontent.com/32401852/61795574-fac31000-ae38-11e9-8072-f1563550869c.png)

It no longer genrates noise but still far from desired output because it is not working effectively as described as a low-pass filter. Obviously we can see halftone patterns in image.

We track this issue in a separate issue #10
